pkg/ring/ring.go

```diff
@@ -588,3 +658,45 @@ func (r *Ring) HasInstance(instanceID string) bool {
 	_, ok := instances[instanceID]
 	return ok
 }
+
+func (r *Ring) getCachedShuffledSubring(identifier string, size int) *Ring {
+	r.mtx.RLock()
+	defer r.mtx.RUnlock()
+
+	if r.lastTopologyChange.IsZero() || r.lastRingChange.IsZero() {
+		// Don't do any caching if this ring has no updates yet.
+		return nil
+	}
+
+	cached := r.shuffledSubringCache[identifier]
+	if cached == nil {
+		return nil
+	}
+
+	cached.mtx.Lock()
+	defer cached.mtx.Unlock()
+
+	if cached.ringShardSize != size {
+		return nil
+	}
+
+	if !cached.lastTopologyChange.Equal(r.lastTopologyChange) {
+		// ingesters changed since cached ring was created, cannot be reused.
+		return nil
+	}
+
+	if cached.lastRingChange.Equal(r.lastRingChange) {
```

Comment:
While it's true that ring changes continuously, distributor limits its updates to one per second (by default, when using Consul). I think it still makes sense to avoid doing extra work here, especially as it increases with shard size.

Ran benchmarks (with no ring updates, but each test was only running for one second anyway), with the if-check, and without.

```
Ring_ShuffleShardCached/num_instances_=_50,_num_zones_=_1,_shard_size_=_3-4       85.7ns ± 3%   230.0ns ± 4%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_50,_num_zones_=_1,_shard_size_=_10-4      83.2ns ± 1%   553.3ns ± 1%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_50,_num_zones_=_1,_shard_size_=_30-4      85.2ns ± 3%  1557.3ns ± 4%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_50,_num_zones_=_3,_shard_size_=_3-4       84.8ns ± 2%   225.0ns ± 1%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_50,_num_zones_=_3,_shard_size_=_10-4      82.4ns ± 0%   637.0ns ± 3%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_50,_num_zones_=_3,_shard_size_=_30-4      83.0ns ± 0%  1525.3ns ± 2%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_100,_num_zones_=_1,_shard_size_=_3-4      83.1ns ± 1%   220.0ns ± 3%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_100,_num_zones_=_1,_shard_size_=_10-4     84.1ns ± 1%   550.3ns ± 2%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_100,_num_zones_=_1,_shard_size_=_30-4     82.6ns ± 1%  1524.3ns ± 3%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_100,_num_zones_=_3,_shard_size_=_3-4      81.7ns ± 1%   219.7ns ± 3%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_100,_num_zones_=_3,_shard_size_=_10-4     82.0ns ± 1%   644.7ns ± 1%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_100,_num_zones_=_3,_shard_size_=_30-4     81.5ns ± 1%  1518.3ns ± 4%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_1000,_num_zones_=_1,_shard_size_=_3-4     82.5ns ± 1%   219.0ns ± 1%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_1000,_num_zones_=_1,_shard_size_=_10-4    81.4ns ± 2%   549.0ns ± 5%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_1000,_num_zones_=_1,_shard_size_=_30-4    82.3ns ± 1%  1493.0ns ± 2%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_1000,_num_zones_=_3,_shard_size_=_3-4     82.7ns ± 1%   219.3ns ± 1%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_1000,_num_zones_=_3,_shard_size_=_10-4    81.5ns ± 2%   645.0ns ± 6%   ~     (p=0.100 n=3+3)
Ring_ShuffleShardCached/num_instances_=_1000,_num_zones_=_3,_shard_size_=_30-4    80.8ns ± 1%  1490.3ns ± 1%   ~     (p=0.100 n=3+3)
```

Reply:
Given this benchmark is on the `ShuffleShard()` function only (which is a tiny part of the whole ingestion path in distributors) and the benefit is few %, I've the feeling the result enforces my feedback to simplify the logic :)

Reply:
> and the benefit is few %

I wouldn't call ~80ns -> 1500ns jump a few % :) But in the context of entire ingestion path, it is indeed tiny fraction of it. Let's simplify.